File: the-method-of-programming/ch01-string/03-all-permutation/exercise-01.js
'use strict'

function convertBase(num, base, len) {
  const digits = []

  while (num > 0) {
    const r = num % base
    digits.unshift(r)
    num = (num - r) / base
  }

  while (digits.length < len) {
    digits.unshift(0)
  }

  return digits
}

function calcAllPermutations(str) {
  const base = str.length
  const max = base ** base
  let i = 0

  while (i < max) {
    const digits = convertBase(i, base, base)
    const permutation = digits.map(d => str[d]).join('')
    console.log(permutation)
    i++
  }
}

function main() {
  calcAllPermutations('abc')
}

main()

File: the-method-of-programming/ch01-string/03-all-permutations/exercise-01.cpp
#include <cstring>
#include <iostream>
#include <cmath>

void calcAllPermutations(const char *str, const int len) {
  int n = 0;
  const int max = pow(len, len);
  char *digits = new char[len];

  while (n < max) {
    int i = n;
    int j = 0;

    while (i > 0) {
      const int r = i % len;
      digits[len - j - 1] = str[r];
      i = (i - r) / len;
      ++j;
    }

    while (j < len) {
      digits[len - j - 1] = str[0];
      ++j;
    }

    std::cout << digits << " ";
    ++n;
  }

  std::cout << std::endl;
}

int main() {
  char str[] = "abc";
  calcAllPermutations(str, strlen(str));

  return 0;
}
